fix(sdk) depositETH gas estimation bug

tsite opened the pull request:
**Description**

Fix a bug in the sdk where gas estimation uses the zero address as the source instead of propagating the source address for the transaction.

**Tests**

Depositing eth via the CrossChainMessenger now works properly and estimateGas no longer reverts due to a lack of funds available at the zero address.

**Metadata**

- Fixes #6723 #7896

tsite committed:
fix issues 6723, 7896

propagate the deposit address instead of using the zero address
as the source for eth_estimateGas

roninjin10 commented:
I believe this code has tests and I'm not 100% clear on why gasEstimation would fail here. Can you please write test that fails before your change demonstrating the issue?

roninjin10 reviewed (changes requested):
I believe this code has tests and I'm not 100% clear on why gasEstimation would fail here. Can you please write test that fails before your change demonstrating the issue?

roninjin10 commented on packages/sdk/src/cross-chain-messenger.ts:
```diff
@@ -2252,20 +2252,33 @@ export class CrossChainMessenger {
         }
-        const gasEstimation = await this.estimateGas.depositETH(amount, opts)
+        // if we don't include the users address the estimation may fail
+        const from =
+          opts?.overrides?.from ??
+          (ethers.Signer.isSigner(this.l1SignerOrProvider)
+            ? (this.l1SignerOrProvider as Signer).getAddress()
+            : undefined)
```
Does ethers not already default to the signer address?

tsite replied:
it was using the provider instead of signer for gas estimation - I updated the fix to use l1SignerOrProvider instead of l1Provider and cleaned up depositERC20 as well.

tsite replied:
@roninjin10 if you could take another look at the pr, that would be awesome

roninjin10 replied:
See my comment here https://github.com/ethereum-optimism/optimism/pull/9949#pullrequestreview-1953937195

tsite replied:
ok I updated the depositETH test to catch regressions @roninjin10 - the tests now fail without the changes made in this pr

tsite committed:
cleanup the fix

we need to use the signer instead of provider to estimate the gas properly

tsite committed:
fix all estimateGas calls

use the signer instead of the provider

tsite committed:
pnpm lint:fix

tsite committed:
update tests to catch regressions

tsite committed:
rerun flakey tests (empty commit)

roninjin10 commented on packages/sdk/test/cross-chain-messenger.spec.ts:
```diff
@@ -25,6 +25,22 @@ describe('CrossChainMessenger', () => {
   let l2Signer: any
   before(async () => {
     ;[l1Signer, l2Signer] = await ethers.getSigners()
+    // ethers.getSigners() is out of date - recent versions of geth
+    // include a balance check as part of the estimateGas() call.
```
One issue with this is optimism contracts themselves expect you to estimate gas with Address(1) to correctly estimate https://github.com/ethereum-optimism/optimism/blob/5bafddac62e41a0b6744d0aac5a0a178b11788be/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L381

tsite replied:
that is okay since the function is non-payable

roninjin10 commented on packages/sdk/test/cross-chain-messenger.spec.ts:
```diff
@@ -25,6 +25,22 @@ describe('CrossChainMessenger', () => {
+    // include a balance check as part of the estimateGas() call.
+    for (const signer of [l1Signer, l2Signer]) {
+      const estimateGas = signer.provider.estimateGas
+      signer.provider.estimateGas = async function (tx) {
```
So if I'm understanding this correctly, the issue here is geth reverts on balance checks. What I'm failing to understand though is we already have these tests running against geth via our devnet tests and against anvil via our anvil/vitest tests

tsite replied:
I think new versions of geth do fail the gas estimate if the balance is lower than the call value. I'll update the test to check for strict inequality since a zero balance is permitted for non-payable functions.

tsite replied:
for example, cast rpc eth_estimateGas '{"to":"0x00a149B262EA8ACA092Af1c43a899e7B18C5A4De", "from":"0x00a149B262EA8ACA092Af1c43a899e7B18C5A4De","value":"0x1"}' fails on most networks due to insufficient balance while cast rpc eth_estimateGas '{"to":"0x00a149B262EA8ACA092Af1c43a899e7B18C5A4De", "from":"0x00a149B262EA8ACA092Af1c43a899e7B18C5A4De","value":"0x0"}' succeeds.

roninjin10 replied:
hmmm definitely confusing to me because I'm pretty sure we are using a new version of geth for our devnet tests. I think you are right though and I just don't know enough to be able to approve this pr. @smartcontracts will check it out within a day or 2. Thank you for patience.

roninjin10 reviewed (commented):
Thanks for adding the tests and apologies if getting this pr merged has been fustrating. My general inclination is to not accept this change but I'm borderline. 

@smartcontracts any thoughts? I will defer to you if you think this change is worth merging. As far as I can tell it doesn't hurt anything

tsite committed:
fix balance check in tests

The sender is allowed to have no balance when calling estimateGas on
nonpayable functions.

Also revert some unnecessary test changes

tsite commented:
I updated the test - would appreciate another review. We currently have patched code running in production to fix this bug and it would be nice to merge the fix upstream as well
